### PR TITLE
Tighten rules for DSVIN/DSTX

### PR DESCRIPTION
**Pull request by UdjinM6**

### CODE REVIEW

File: src/privatesend-server.cpp [hunk 1]
```diff
@@ -171,15 +171,28 @@ void CPrivateSendServer::ProcessMessage(CNode* pfrom, const std::string& strComm
 
         LogPrint("privatesend", "DSVIN -- txCollateral %s", entry.txCollateral->ToString());
 
+        auto consumeCollateral = [&](const CTransactionRef& txref) {
+            LOCK(cs_main);
+            CValidationState validationState;
+            if (!AcceptToMemoryPool(mempool, validationState, txref, true, NULL)) {
```
Comment (nmarley) [inline src/privatesend-server.cpp]: nit: `nullptr` instead?

**Comment by nmarley** [comment]
Just curious -- any specific reason `consumeCollateral` is a lambda instead of a regular function? Or just simpler since it's only called in `CPrivateSendServer::ProcessMessage` scope?

**Comment by UdjinM6** [comment]
@nmarley That's a good point, we can reuse it in a couple of other places actually. Fixed.

File: src/privatesend-server.cpp [hunk 2]
```diff
@@ -174,12 +174,14 @@ void CPrivateSendServer::ProcessMessage(CNode* pfrom, const std::string& strComm
...
+            ConsumeCollateral(connman, entry.txCollateral);
             return;
         }
 
         if (entry.vecTxOut.size() > PRIVATESEND_ENTRY_MAX_SIZE) {
             LogPrintf("DSVIN -- ERROR: too many outputs! %d/%d\n", entry.vecTxOut.size(), PRIVATESEND_ENTRY_MAX_SIZE);
             PushStatus(pfrom, STATUS_REJECTED, ERR_MAXIMUM, connman);
+            ConsumeCollateral(connman, entry.txCollateral);
             return;
         }
```
Comment (PastaPastaPasta) [inline src/privatesend-server.cpp]: Do we never check if vecTxDSIn == vecTxDSOut?

Comment (UdjinM6) [reply src/privatesend-server.cpp]: Good point (assuming you were talking about their size) - fc853b176

Comment (PastaPastaPasta) [reply src/privatesend-server.cpp]: based on that, doing this check is kindof redundant, if they are equal there's no need to check both

Comment (thephez) [reply src/privatesend-server.cpp]: Re: max size check? There's still a max size limit. You could have equal numbers of inputs / outputs, but have too many of them (e.g. 50 inputs / 50 outputs).

File: src/privatesend.cpp [hunk 3]
```diff
@@ -122,6 +122,26 @@ bool CPrivateSendBroadcastTx::IsExpired(int nHeight)
...
 
+bool CPrivateSendBroadcastTx::IsValidStructure()
+{
+    // some trivial checks only
+    if (tx->vin.size() != tx->vout.size()) {
+        return false;
+    }
+    if (tx->vin.size() < CPrivateSend::GetMinPoolParticipants()) {
+        return false;
+    }
```
Comment (PastaPastaPasta) [inline src/privatesend.cpp]: check if greater than GetMaxPoolParticipants

Comment (UdjinM6) [reply src/privatesend.cpp]: fd8a5a3

**Comment by UdjinM6** [comment]
Rebased to fix log category after #2918

**Pull request merged**